### ISSUE TITLE
document branch and release strategy

### ISSUE THREAD

**Issue by busbey** [brianfrankcooper/YCSB#244]
We need to decide how often we'll release and how we'll maintain different release versions.

**Comment by busbey** [comment]
Unless someone objects, I'm going to aim for a the following:

monthly release cadence. If it turns out there's a bunch of changes after the backlog clears, we can go for every 2 weeks.

My plan is to only maintain master as a long lived branch.

When a change comes in that requires a minor or major version bump I'll establish a temporary branch until the current in-progress release goes out. After, that branch will get merged into Master and that new minor/major release will go out.

We'll only do updates to older minor/major release lines when someone asks. In that on-demand case we'll branch based on the last release tag and then cherry pick the needed additions.

**Comment by gkamat** [comment]
That sounds like a reasonable plan.

**Comment by busbey** [comment]
I've been thinking about this more in anticipation of starting the 0.2.0 release process. I think we should do the following instead:

* master is only long-lived branch
* master-first policy (changes go to master before anywhere else, unless they are only applicable to a since-obviated part of the code)
* when changes need to be applied to an older release line, cherry-pick
* de-facto feature branches in the forks of other contributors
* use release branches temporarily

by temporarily, I mean we'll start a release branch under these conditions:

* start of the release candidate process (so that changes unrelated to polishing the release can still go to master)
* availability of a change that requires a version bump (per our Version Numbers and Compatibility guide)

The first is so that we can declare "we're starting the release for version X" and then cut provisional releases off of that branch when folks find something wrong. If there's a new feature or a non-critical fix that might introduce new problems, it goes to master. If there's a fix that we want to include in the release, it goes to master first and then gets cherry-picked.

The nice part about this is that it means doing a patch release for an older version number is basically the same, except the initial branch starts from the previous release tag instead of then-current master.

The second branch criteria is so that we can still quickly accept changes that break compatibility. Essentially we just maintain the release line for the previous compatible release a little bit longer than usual (since that branch should have been released at most in ~1 month).

Does this sound reasonable? If so I can get a nicer-written version on a page. It would be better to have an explanation prior to getting 0.2.0 out the door.

**Comment by cmccoy** [comment]
Thanks for writing this up! I like having a single long-lived branch, and
using RC-branches to freeze features.

Could you clarify how patches for older versions will work? Would the
branch off the previous release just last until the patch release is out?
So if the current release is 0.2.0, and previous release was 0.1.4, a
branch (say release-0.1.5) would be created until 0.1.5 is released? Do we
want to commit to back-porting patches?

**Comment by busbey** [comment]
> Could you clarify how patches for older versions will work? Would the
> branch off the previous release just last until the patch release is out?
> So if the current release is 0.2.0, and previous release was 0.1.4, a
> branch (say release-0.1.5) would be created until 0.1.5 is released? Do we
> want to commit to back-porting patches?

Yes, that's what I'm suggesting. We'd create a 0.1.5 branch off of the 0.1.4 tag and then just keep it until we released 0.1.5. If we try to stick to having a "release window" in a given month, that could mean sticking around for other fixes until the release. Alternatively, we could immediately start a release candidate that presumably wouldn't take long (since changes would be small). Then the branch would go away just like our normal release branches.

I don't think we need a stated policy on back-porting patches. I figure as things get more established we can do things on a case-by-case basis until we have more information about how long older versions are needed. I figure if someone is working towards e.g. publishing some benchmarking and runs into a problem, they can mention the version they need in their problem report and we can cut them a patch release so they can still compare results to what they've already measured. (presuming the fix in question can be done in a compatible way.)

**Comment by cmccoy** [comment]
Sounds good to me.
